File: Projects/Compiladores/trabalho1/Main.cpp
#include <iostream>
#include <string>
using namespace std;

// Compiladores trabalho 01 - implementação AFD análise léxica em C++

int main() {

    string entrada = "";
    //std::cout << "Digite a entrada: ";
    //std::getline(std::cin, entrada);
    
    int i = 0;
    int estado = 0;
    char caractere_entrada;
    std::string Token = "";

    bool flag = true;
    bool erro = false;

    while (flag == true){

        std::cout << "\n O que deseja fazer?" << std::endl;
        std::cout << "1 - Ler a entrada" << std::endl;
        std::cout << "2 - Sair" << std::endl;
        int opcao;
        std::cin >> opcao;

        switch (opcao) {
            case 1:

                std::cout << "Digite a entrada: ";
                std::cin.ignore(); // Limpa o buffer
                std::getline(std::cin, entrada);
                
                while (estado == 0 || estado == 1 || estado == 2 || estado == 3 || estado == 4) {
        
                    if (i >= entrada.size()) {
            
                        // Fim da string, sair do loop
                        estado = 5;
                        break;
                    }
            
                    caractere_entrada = entrada[i];
                    Token += caractere_entrada;
                    i++;
            
                    switch (estado) {
                        case 0:
                            std::cout << "\n Estado 0" << std::endl;
                            std::cout << "Caractere lido: " << caractere_entrada << std::endl;
                            
                            if (isdigit(caractere_entrada)) {
                                std::cout << "Um Dígito foi lido" << std::endl;
                                estado = 2;
                            } else if (caractere_entrada == 32) {
                                std::cout << "Um espaço em branco foi lido" << std::endl;
                                estado = 0;
                            } else if (caractere_entrada == '+' || caractere_entrada == '-') {
                                std::cout << "Um Sinal foi lido" << std::endl;
                                estado = 1;
                            } else {
                                std::cout << "Um Caractere invalido foi lido" << std::endl;
                                estado = -1;
                                erro = true;
                            }
            
                            std::cout << "\n Token lido: " << Token << std::endl;
            
                            break;
            
                        case 1:
                            std::cout << "\n Estado 1" << std::endl;
                            std::cout << "Caractere lido: " << caractere_entrada << std::endl;

                            if (isdigit(caractere_entrada)) {
                                std::cout << "Um Dígito foi lido" << std::endl;
                                estado = 2;
                            } else {
                                std::cout << "Um Caractere invalido foi lido" << std::endl;
                                estado = -1;
                                erro = true;
                            }
                            std::cout << "\n Token lido: " << Token << std::endl;
            
                            break;
            
                        case 2:
                            std::cout << "\n Estado 2" << std::endl;
                            std::cout << "Caractere lido: " << caractere_entrada << std::endl;

                            if (isdigit(caractere_entrada)) {
                                std::cout << "Um Dígito foi lido" << std::endl;
                                estado = 2;
                            } else if (caractere_entrada == ',') {
                                std::cout << "Uma virgula foi lida" << std::endl;
                                estado = 3;
                            } else {
                                std::cout << "Um outro caractere foi lido" << std::endl;
                                estado = 5;
                                erro = true;
                            }
            
                            std::cout << "\n Token lido: " << Token << std::endl;
            
                            break;
            
                        case 3:
                            std::cout << "\n Estado 3" << std::endl;
                            std::cout << "Caractere lido: " << caractere_entrada << std::endl;

                            if (isdigit(caractere_entrada)) {
                                std::cout << "Um Dígito foi lido" << std::endl;
                                estado = 4;
                            } else {
                                std::cout << "Um Caractere invalido foi lido" << std::endl;
                                estado = -1;
                                erro = true;
                            }
            
                            std::cout << "\n Token lido: " << Token << std::endl;
            
                            break;
            
                        case 4:
                            std::cout << "\n Estado 4" << std::endl;
                            std::cout << "Caractere lido: " << caractere_entrada << std::endl;

                            if (isdigit(caractere_entrada)) {
                                std::cout << "Um Dígito foi lido" << std::endl;
                                estado = 4;
                            } else {
                                std::cout << "Um outro caractere foi lido" << std::endl;
                                estado = 5;
                            }
            
                            std::cout << "\n Token lido: " << Token << std::endl;
            
                            break;
                    }
                } 

                if (estado == 5 && erro == false) {
                    std::cout << "\n Entrada aceita" << std::endl;
                    estado = 0;
                    i = 0;
                    
                } else {
                    std::cout << "\n Erro de leitura" << std::endl;
                }
            
                std::cout << "\n Token final: " << Token << std::endl;
                Token = "";
            
                break;
            case 2:
                std::cout << "Saindo..." << std::endl;
                flag = false;
                break;
            default:
                std::cout << "Opção inválida. Tente novamente." << std::endl;
        }

           
    }
    
    return 0;
}

File: Projects/Compiladores/trabalho1/Main.py
def analisador_lexico_tipo2(entrada):
    # Definição dos estados
    INICIAL = 0      # Estado inicial
    INTEIRO = 1      # Encontrou dígito(s)
    VIRGULA = 2      # Encontrou vírgula após dígito
    DECIMAL = 3      # Encontrou dígito(s) após vírgula
    
    # Estados finais
    estados_finais = [INTEIRO, DECIMAL]
    
    # Variável global para o estado corrente (conforme a implementação Tipo 2)
    estado_corrente = INICIAL
    
    # Índice atual na string de entrada
    i = 0
    
    # Lexema reconhecido
    lexema = ""
    
    # Laço principal do AFD
    while i < len(entrada):
        # Caractere atual
        c = entrada[i]
        
        # Switch para controlar a transição de estados
        if estado_corrente == INICIAL:
            if c.isdigit():
                lexema += c
                estado_corrente = INTEIRO
            else:
                # Ignoramos caracteres que não iniciam um número
                i += 1
                continue
                
        elif estado_corrente == INTEIRO:
            if c.isdigit():
                lexema += c
            elif c == ',':
                lexema += c
                estado_corrente = VIRGULA
            elif c == '.':
                # Encontrou ponto, então retorna apenas o inteiro
                print(f"Token reconhecido: {lexema}")
                return lexema
            else:
                # Qualquer outro caractere termina o reconhecimento
                print(f"Token reconhecido: {lexema}")
                return lexema
                
        elif estado_corrente == VIRGULA:
            if c.isdigit():
                lexema += c
                estado_corrente = DECIMAL
            else:
                # Formato inválido após a vírgula
                print("Erro: formato inválido após a vírgula")
                return None
                
        elif estado_corrente == DECIMAL:
            if c.isdigit():
                lexema += c
            else:
                # Qualquer outro caractere termina o reconhecimento
                print(f"Token reconhecido: {lexema}")
                return lexema
        
        # Avança para o próximo caractere
        i += 1
        
        # Se chegamos ao final da entrada e estamos em um estado final, aceitamos
        if i >= len(entrada) and estado_corrente in estados_finais:
            print(f"Token reconhecido: {lexema}")
            return lexema
    
    # Se chegamos aqui e não estamos em um estado final, rejeitamos
    if estado_corrente not in estados_finais:
        print("Nenhum token foi reconhecido")
        return None
    else:
        print(f"Token reconhecido: {lexema}")
        return lexema


def analisador_lexico_tipo2_c(entrada):
    """
    Versão em pseudocódigo C da implementação Tipo 2
    """
    print("Versão equivalente em C:")
    codigo_c = """
    #include <stdio.h>
    #include <string.h>
    #include <ctype.h>

    char* analisador_lexico(const char* entrada) {
        // Definição dos estados
        enum Estados {
            INICIAL = 0,      // Estado inicial
            INTEIRO = 1,      // Encontrou dígito(s)
            VIRGULA = 2,      // Encontrou vírgula após dígito
            DECIMAL = 3       // Encontrou dígito(s) após vírgula
        };
        
        // Variável global para o estado corrente (conforme a implementação Tipo 2)
        int estado_corrente = INICIAL;
        
        // Índice atual na string de entrada
        int i = 0;
        
        // Lexema reconhecido
        char lexema[100] = "";
        int lexema_idx = 0;
        
        // Laço principal do AFD
        while (entrada[i] != '\\0') {
            // Caractere atual
            char c = entrada[i];
            
            // Switch para controlar a transição de estados
            switch (estado_corrente) {
                case INICIAL:
                    if (isdigit(c)) {
                        lexema[lexema_idx++] = c;
                        estado_corrente = INTEIRO;
                    } else {
                        // Ignoramos caracteres que não iniciam um número
                        i++;
                        continue;
                    }
                    break;
                    
                case INTEIRO:
                    if (isdigit(c)) {
                        lexema[lexema_idx++] = c;
                    } else if (c == ',') {
                        lexema[lexema_idx++] = c;
                        estado_corrente = VIRGULA;
                    } else if (c == '.') {
                        // Encontrou ponto, então retorna apenas o inteiro
                        lexema[lexema_idx] = '\\0';
                        printf("Token reconhecido: %s\\n", lexema);
                        return strdup(lexema);
                    } else {
                        // Qualquer outro caractere termina o reconhecimento
                        lexema[lexema_idx] = '\\0';
                        printf("Token reconhecido: %s\\n", lexema);
                        return strdup(lexema);
                    }
                    break;
                    
                case VIRGULA:
                    if (isdigit(c)) {
                        lexema[lexema_idx++] = c;
                        estado_corrente = DECIMAL;
                    } else {
                        // Formato inválido após a vírgula
                        printf("Erro: formato inválido após a vírgula\\n");
                        return NULL;
                    }
                    break;
                    
                case DECIMAL:
                    if (isdigit(c)) {
                        lexema[lexema_idx++] = c;
                    } else {
                        // Qualquer outro caractere termina o reconhecimento
                        lexema[lexema_idx] = '\\0';
                        printf("Token reconhecido: %s\\n", lexema);
                        return strdup(lexema);
                    }
                    break;
            }
            
            // Avança para o próximo caractere
            i++;
        }
        
        // Se chegamos ao final da entrada e estamos em um estado final, aceitamos
        lexema[lexema_idx] = '\\0';
        if (estado_corrente == INTEIRO || estado_corrente == DECIMAL) {
            printf("Token reconhecido: %s\\n", lexema);
            return strdup(lexema);
        } else {
            printf("Nenhum token foi reconhecido\\n");
            return NULL;
        }
    }

    int main() {
        const char* testes[] = {
            "var := 3,14",  // Deve reconhecer 3,14
            "var := 3.14",  // Deve reconhecer apenas 3
            "123,456",      // Deve reconhecer 123,456
            "123.",         // Deve reconhecer 123
            "45",           // Deve reconhecer 45
            "var := abc",   // Não deve reconhecer nenhum token
            ",123",         // Não deve reconhecer (não começa com dígito)
            "123,",         // Erro: termina com vírgula
            "0,005"         // Deve reconhecer 0,005
        };
        
        for (int i = 0; i < 9; i++) {
            printf("\\nEntrada: '%s'\\n", testes[i]);
            char* resultado = analisador_lexico(testes[i]);
            printf("Resultado: %s\\n", resultado ? resultado : "NULL");
            if (resultado) free(resultado);
        }
        
        return 0;
    }
    """
    print(codigo_c)
    return "Código C fornecido acima"


# Função de teste para a implementação em Python
def testar_analisador():
    testes = [
        "var := 3,14",     # Deve reconhecer 3,14
        "var := 3.14",     # Deve reconhecer apenas 3
        "123,456",         # Deve reconhecer 123,456
        "123.",            # Deve reconhecer 123
        "45",              # Deve reconhecer 45
        "var := abc",      # Não deve reconhecer nenhum token
        ",123",            # Não deve reconhecer (não começa com dígito)
        "123,",            # Erro: termina com vírgula
        "0,005"            # Deve reconhecer 0,005
    ]
    
    print("Testes da implementação Tipo 2 em Python:")
    for teste in testes:
        print(f"\nEntrada: '{teste}'")
        resultado = analisador_lexico_tipo2(teste)
        print(f"Resultado: {resultado}")
    
    # Mostrar o código em C
    print("\n" + "="*50 + "\n")
    analisador_lexico_tipo2_c("")


if __name__ == "__main__":
    testar_analisador()